### PR TITLE
Scala sh rewrite

### PR DESCRIPTION
**Pull request by BarkingBad**

### CODE REVIEW

**Comment by BarkingBad** [comment]
Does anyone know why we couldn't create val of type `Regex` then reuse it and we had to inline the regex in the if condition? My last commit changes this and then the CI passes. Previously it was failing with for `CompilationTests.tastyBootstrap`:

```
[info] Test dotty.tools.dotc.CompilationTests.tastyBootstrap started
possible data race involving globally reachable variable root in class Pattern: java.util.regex.Pattern#Node
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable matchRoot in class Pattern: java.util.regex.Pattern#Node
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable buffer in class Pattern: Array[Int]
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable predicate in class Pattern: java.util.regex.Pattern#CharPredicate
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable namedGroups in class Pattern: java.util.Map[String, Integer]
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable groupNodes in class Pattern: Array[java.util.regex.Pattern#GroupHead]
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable topClosureNodes in class Pattern: java.util.List[java.util.regex.Pattern#Node]
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable localTCNCount in class Pattern: Int
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable hasGroupRef in class Pattern: Boolean
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable capturingGroupCount in class Pattern: Int
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
possible data race involving globally reachable variable localCount in class Pattern: Int
  use -Ylog:checkReentrant+ to find out more about why the variable is reachable.
Compilation failed for: 'dotty1 from tastyBootstrap/dotty1'                     
Error:  Test dotty.tools.dotc.CompilationTests.tastyBootstrap failed: java.lang.AssertionError: Expected no errors when compiling, failed for the following reason(s):
Error:  
Error:    - encountered 1 test failures(s), took 34.712 sec
Error:      at dotty.tools.vulpix.ParallelTesting$CompilationTest.checkCompile(ParallelTesting.scala:897)
Error:      at dotty.tools.dotc.CompilationTests.$anonfun$2(CompilationTests.scala:288)
Error:      at scala.collection.immutable.List.map(List.scala:250)
Error:      at dotty.tools.dotc.CompilationTests.tastyBootstrap(CompilationTests.scala:288)
Error:      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Error:      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
Error:      at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Error:      at java.lang.reflect.Method.invoke(Method.java:567)
Error:      ...
```

Also, for coursier tests I need artifacts to be published. Is it OK @smarter or I should redesign it, for example move it to the separate CI action?

**Comment by smarter** [comment]
> Does anyone know why we couldn't create val of type Regex 

The analysis done by CheckReentrant is imprecise and incorrectly thinks that java regexps are not thread-safe, but it's actually fine to use them, so we just turn off the check using `@sharable`: https://github.com/lampepfl/dotty/blob/839ef3203e654a38b431558eea52821c974a1734/compiler/src/dotty/tools/repl/ParseResult.scala#L114 (though we could also special-case regexps in CheckReentrant)

> Also, for coursier tests I need artifacts to be published. Is it OK @smarter or I should redesign it

What's important is that if I run `test` locally, all tests should succeed, so we can't have tests enabled by default that require some extra step.

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+  @tailrec
+  def process(args: List[String], settings: Settings): Settings = args match
+    case Nil =>
+      settings
+    case "-repl" :: tail =>
+      process(tail, settings.withExecuteMode(ExecuteMode.Repl))
+    case "-run" :: tail =>
+      process(tail, settings.withExecuteMode(ExecuteMode.Run))
+    case ("-cp" | "-classpath" | "--classpath") :: cp :: tail =>
```
Comment (smarter) [inline compiler/src/dotty/tools/MainGenericRunner.scala:57]: The long option is called --class-path, not --classpath

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  end withTargetScript
+}
+
+object MainGenericRunner {
+
+  @tailrec
+  def process(args: List[String], settings: Settings): Settings = args match
+    case Nil =>
+      settings
+    case "-repl" :: tail =>
```
Comment (smarter) [inline compiler/src/dotty/tools/MainGenericRunner.scala:53]: In which situation is it necessary to explicitly pass -repl?

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  def withTargetScript(file: String): Settings =
+    Try(Source.fromFile(file)).toOption match
+      case Some(_) => this.copy(targetScript = file)
+      case None      =>
+        println(s"not found $file")
+        this.copy(exitCode = 2)
+  end withTargetScript
+}
+
+object MainGenericRunner {
```
Comment (smarter) [inline compiler/src/dotty/tools/MainGenericRunner.scala:47]: If this is supposed to be the main entry point for command line users, the various bash scripts and batch scripts in dist/bin should be adapted to call into it and not duplicate logic with it.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 1 above)
Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:66]: Will change

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 2 above)
Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:62]: Actually, never. But `dist/bin` script provides such option, so we wanted to recreate this

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 3 above)
Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:54]: That is right. Will change them

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 2 above)
Comment (smarter) [reply compiler/src/dotty/tools/MainGenericRunner.scala:62]: If it's never used let's just remove it then.

**Comment by smarter** [comment]
The last commit committed .class and .tasty files that shouldn't be in git.

**Comment by BarkingBad** [comment]
> The last commit committed .class and .tasty files that shouldn't be in git.

I did it on purpose for testing `-run` but I can recompile the `.scala` file indeed.

File: compiler/src/dotty/tools/MainGenericRunner.scala (hunk 2 above)
Comment (philwalk) [reply compiler/src/dotty/tools/MainGenericRunner.scala:62]: > In which situation is it necessary to explicitly pass -repl?

In `dist/bin/scalac`, it's needed, but doesn't seem to be needed in `dist/bin/scala`.

**Comment by lrytz** [comment]
This PR addresses https://github.com/coursier/coursier/issues/2058, right?

**Comment by BarkingBad** [comment]
@lrytz  I was not aware about this issue. However I am afraid it will take some time till we get it released to maven central, so the coursier descriptor can be created for `scala3` as it will be release  with  `3.0.3`

**Comment by smarter** [comment]
if this is deemed important enough, we could backport it to the 3.0.2 branch.

**Comment by philwalk** [comment]
@BarkingBad -  I like the idea of migrating as much functionality from sh script to scala code, although I'm not entirely clear as to the scope of this PR.  Will there be support for `jline` console functionality?

**Comment by BarkingBad** [comment]
What do you mean @philwalk? If you mean that coursier tests cannot initialize jline I think it's the internal problem from calling in the virtual terminal. Calling REPL works just fine for both shell script and when we eventually turn it into coursier executable
obraz

**Comment by philwalk** [comment]
> What do you mean @philwalk?

I'm not referring specifically to the coursier use-case, but to whether and how a Windows `cygwin` or `msys` environment (for example) might be affected by this PR.  Possibly I just need to understand some context.

When I launch a REPL session from a `cygwin` bash terminal, here's what I see with these changes:
```sh
# philwalk@d5 ~/workspace/dotty-sh-rewrite
$ bin/scala
Starting scala3 REPL...
←[?1h←=←[90m~←[0m
←[?2004h←[34mscala> ←[0m:q
:q
←[?1l←>←[?1000l←[?2004l
```
By contrast, here's what happens with the current scala release:
```sh
$ /opt/scala3/bin/scala
scala> :q
```
Also, there is other lost functionality compared

**Comment by BarkingBad** [comment]
I had no idea about cygwin issues, if you would like to contribute feel free. I had hard time deciphering bash scripts (some things seemed to be unused or redundant) so I could omit some functionality, for example I don't know what `SCALA_OPTS` are for

**Comment by philwalk** [comment]
> I had no idea about cygwin issues, if you would like to contribute feel free. I had hard time deciphering bash scripts (some things seemed to be unused or redundant) so I could omit some functionality, for example I don't know what `SCALA_OPTS` are for.
> 
> EDIT: For cygwin issues, I actually cannot test it since I don't have any windows machine available at the moment, I'd appreciate your help in that matter

I'm happy to contribute.

The purpose of `SCALA_OPTS` is to provide a way to specify default arguments in a scripting and/or a REPL environment, similar to the way `JAVA_OPTS` works. 
An example for how it could be configured in a SHELL environment:
```sh
SCALA_OPTS="@/home/philwalk/.myScalaClasspath -save -deprecation -color:never"
```
The most important use case is to permit an externally managed custom classpath, thereby enabling support for simple but non-trivial scala scripts, e.g.:
```bash
#!/opt/scala3/bin/scala
import better.files._
import import scalikejdbc._
def main(args: Array[String]): Unit = ???
```

Obviously, it can also be used to set other defaults arguments as well.

To support it in MainGenericRunner, here's a simple whitespace-delimited interpretation of `SCALA_OPTS` that inserts default arguments ahead of `args`. 

```scala
import dotty.tools.dotc.config.Properties._

  lazy val scalaOpts = envOrElse("SCALA_OPTS","").split(" ")

  def main(actualargs: Array[String]): Unit =
      val args = scalaOpts ++ actualargs
      val settings = process(args.toList, Settings())

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        val properArgs =
+          List("-classpath", settings.classPath.mkString(classpathSeparator)).filter(Function.const(settings.classPath.nonEmpty))
+            ++ settings.residualArgs
+        repl.Main.main(properArgs.toArray)
+      case ExecuteMode.Run =>
+        val properArgs =
+          val newClasspath = settings.classPath ++ getClasspath :+ "."
+          List("-classpath", newClasspath.mkString(classpathSeparator)).filter(Function.const(newClasspath.nonEmpty))
+            ++ settings.residualArgs
+        s"java ${settings.javaArgs.mkString(" ")} ${properArgs.mkString(" ")}".! // For now we collect classpath that coursier provides for convenience
```
Comment (dwijnand) [inline compiler/src/dotty/tools/MainGenericRunner.scala:124]: No, that's not what we want to do here: we don't want to fork another JVM - we want to launch the main in a new classloader constructed with the given classpath.  In scala/scala that's `ObjectRunner.runAndCatch` if you want to mirror what's being done there.

Comment (BarkingBad) [reply compiler/src/dotty/tools/MainGenericRunner.scala:124]: Thanks for the advice. I couldn't find proper entry from scala sources and didn't think about launching the main in a new classloader. Will change

**Comment by BarkingBad** [comment]
@philwalk thank you for your snippet and explanation, it was really helpful. As it comes to fixing the cygwin issues, would you like to commit on top of my branch or you have some other idea how to do it?

@dwijnand I copied some sources from scala/scala repo to get things going since IIRC there was no similar class in dotty available to reuse. Could you look at the changes?

**Comment by dwijnand** [comment]
> @dwijnand I copied some sources from scala/scala repo to get things going since IIRC there was no similar class in dotty available to reuse. Could you look at the changes?

I didn't check but I didn't think that much was needed.  It's not doing the forking any more, so that LGTM, thanks.

**Comment by philwalk** [comment]
> @philwalk thank you for your snippet and explanation, it was really helpful. As it comes to fixing the cygwin issues, would you like to commit on top of my branch or you have some other idea how to do it?

I'll try to figure out what's missing or different, though it might take me awhile before I can get to it.

**Comment by philwalk** [comment]
A fix for the cygwin console problem is to copy the following line from `dist/bin/scalac` line 72 to `dist/bin/scala` just before calling java (just before the `eval $JAVACMD`, near line 32)

```sh
[ -z "${ConEmuPID-}" -o -n "${cygwin-}" ] && export MSYSTEM= PWD= # workaround for #12405
```

**Comment by BarkingBad** [comment]
Thank you for your response @philwalk. I will try to setup VM Windows to test it as well when I settle up with other issues

**Comment by philwalk** [comment]
@BarkingBad 
We're actually closer than I thought.    Here are some remaining gaps in the implementation, AFAICT:

The `-Dscript.path=<script-path>` property needs to be set in `dist/bin/scala`, and the script `compiler/test-resources/scripting/scriptPathh.sc` should have caused the build to fail.    I'll see about fixing the test so it works as intended.

The decision as to whether an argument is a `<script-path>` should be generous and accept an argument even if the specified file does not exist;  otherwise, when an interactive user misspells a script path, all of the arguments intended for the script will be sent to the compiler, resulting in incomprehensible error messages, or worse.

The release version of `dist/bin/scala` discards `scala_args` when launching `REPL` mode, since they are intended as the args that go to scala when running in script mode.

**Comment by BarkingBad** [comment]
Hi @philwalk 

I implemented your cygwin fix.

Also referring to your comment https://github.com/lampepfl/dotty/pull/13081#issuecomment-884438142 the stacktrace is from the `Run` branch since I didn't implement `@` arguments, therefore he tried to run the classpath as a compiled file. (As you can see it does not enter `dotty.tools.scriptin.Main.main` in the stacktrace) 

The second comment https://github.com/lampepfl/dotty/pull/13081#issuecomment-885159325 about `script.path` is in my opinion not valid in terms of what we are trying to do, which is not to process args in bash script but in scala entry Main. I added passing additional value to with `scirpt.path` to the `scripting.Main` to set up the value.

I rebased onto your branch with tests (since they are not merged yet) and ran these tests locally - all passed :smile:

**Comment by philwalk** [comment]
@BarkingBad 
> I added setting up the value of `script.path` in the `scripting.Main`
Yes, that's better.   Everthing looks good to me!

**Comment by BarkingBad** [comment]
Could you take a look @philwalk one more time and if everything is OK give approval? I need it to merge the changes

**Review by philwalk — APPROVED** [review]
@BarkingBad - everything looks good, I only have one suggestion:
1. in [CoursierScalaTests.scala](https://github.com/lampepfl/dotty/blob/f3a72f2da4c2d3cf65df6deed2edca168f261a62/compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala), the `scriptTest` only verifies that the property is defined, it could also verify that the value endsWith `scriptPath.sc`.

**Comment by bishabosha** [comment]
can we backport this?

**Comment by BarkingBad** [comment]
I think we could, it would be beneficial for all coursier users I guess

**Comment by dwijnand** [comment]
I'd say no: it's ~500 lines of fresh code, let's leave the backports for the critical bugfixes.

**Comment by SethTisue** [comment]
Hey, this is great! 👏 

I was able to figure out how to use this from Coursier; see  https://github.com/coursier/coursier/issues/2058#issuecomment-896490125 for details

**Pull request merged**